### ISSUE TITLE
Searchable directive with `@auth` generates confusing resolver

### ISSUE THREAD

**Issue by taneba** [aws-amplify/amplify-cli#5833]
**Describe the bug**
I'm not sure if this is a bug, but several issues are posted regarding to this problem.
(related issues: #5795 #5009 #2822)

Basically I faced some problems below when running search query with certain limit.

- it returns incorrect length of items
- It returns wrong length of total count
- it doesn’t return nexttoken even though there’re still data

After investigation, it turned out those were caused by the same reason.

Seeing the implementation, the search resolver filters the results based on the @auth directive, which leads to unexpected result especially when there's no filtered data as a result. to be clear, I explain this in the repro section.

**Amplify CLI Version**
4.31.1

**To Reproduce**
let's say you have a  Bool type with @model, @auth and @searchable

```
type Book 
  @model
  @searchable 
  @auth(
    rules: [
      {allow: owner, ownerField: "owner", operations: [create, update, delete, read]}
    ]
  )
{
  id: ID!
  title: String!
  owner: String
}
```

then you want to retrieve all of the books owned by current user, whose title contains “Harry Potter”, so query will be like

```
searchBooks({
  limit: 5,
  sort: {direction: desc, field: "createdAt"},
  filter: { title: { match: "Harry Potter" } }
})
```

It searches in elastic search like this

```
GET book/doc/_search
{
  "size": 5,
  "sort":  [ { "createdAt": { "order": "desc" }  }] 
  "query": {
     "title": { "match": "Harry Potter" }
  }
}
```

As this query doesn't includes owner field, it'll search every books in elastic search, no matter who own them.

This shouldn’t be a problem as long as the search result only contains current-user’s book.
But what if it contains other users book?

In the next step, res.vtl handles that response, checking if the items belong to their owners.

```
    #set( $items = [] )
    #foreach( $item in $es_items )
      #set( $isLocalOwnerAuthorized = false )
      #set( $allowedOwners0 = $util.defaultIfNull($item.uid, []) )
      #set( $identityValue = $util.defaultIfNull($ctx.identity.claims.get("username"), $util.defaultIfNull($ctx.identity.claims.get("cognito:username"), "___xamznone____")) )
      #if( $util.isList($allowedOwners0) )
        #foreach( $allowedOwner in $allowedOwners0 )
          #if( $allowedOwner == $identityValue )
            #set( $isLocalOwnerAuthorized = true )
          #end
        #end
      #end
      #if( $util.isString($allowedOwners0) )
        #if( $allowedOwners0 == $identityValue )
          #set( $isLocalOwnerAuthorized = true )
        #end
      #end
      #if( ($isLocalDynamicGroupAuthorized == true || $isLocalOwnerAuthorized == true) )
        $util.qr($items.add($item))
      #end
    #end
    #set( $es_items = $items )
```

Here, if result doesn’t contains current owners data, it obviously returns empty data with total count 0, and nextToken is going to be null as well.

So as a result, developers might be confused because there're still matched data.

to solve this issue, I added owner field check to the filter

```
searchBooks({
  limit: 5,
  sort: {direction: desc, field: "createdAt"},
  filter: {owner: {eq: "XXXXXXXXXXXXXXXXXX"}, title: {matchPhrase: "Harry Potter"}}
})
```

It works but it's not intuitive. In addition, there're two similar filtering logic in elastic search and res.vtl which is wasteful.

**Expected behavior**
It might should filter only on the request side (elastic search side) when auth rule is applied.

**Desktop (please complete the following information):**
 - OS: Mac
 - Node Version v12.16.0

**Comment by taneba** [comment]
I updated code example because there's some mistakes.

I want to left a suggestion:
Since filter twice is redundant implementation, amplify-cli should create request-side owner filtering logic. And I'm not sure if it's a appropriate way but it should return error if non current owner id is set on owner field (say, 'access denied' or 'authorization error'). I tried to implement this once but couldn't because toElasticsearchQueryDSL is not open sourced.
